include/rmm/mr/device/detail/free_list.hpp

```diff
@@ -16,153 +16,62 @@
-  }
-
-  /**
-   * @brief Verifies whether this block can be merged to the beginning of block b.
-   *
-   * @param b The block to check for contiguity.
-   * @return true Returns true if this blocks's `ptr` + `size` == `b.ptr`, and `not b.is_head`,
-                  false otherwise.
-   */
-  inline bool is_contiguous_before(block const& b) const noexcept
-  {
-    return (pointer() + size() == b.ptr) and not(b.is_head());
-  }
-
-  /**
-   * @brief Is this block large enough to fit `sz` bytes?
-   *
-   * @param sz The size in bytes to check for fit.
-   * @return true if this block is at least `sz` bytes
-   */
-  inline bool fits(size_t sz) const noexcept { return size() >= sz; }
-
-  /**
-   * @brief Is this block a better fit for `sz` bytes than block `b`?
-   *
-   * @param sz The size in bytes to check for best fit.
-   * @param b The other block to check for fit.
-   * @return true If this block is a tighter fit for `sz` bytes than block `b`.
-   * @return false If this block does not fit `sz` bytes or `b` is a tighter fit.
-   */
-  inline bool is_better_fit(size_t sz, block const& b) const noexcept
-  {
-    return fits(sz) && (size() < b.size() || b.size() < sz);
-  }
-
-  /**
-   * @brief Print this block. For debugging.
-   */
-  void print() const { std::cout << reinterpret_cast<void*>(pointer()) << " " << size() << "B\n"; }
-
- private:
-  char* ptr{};          ///< Raw memory pointer
-  size_t size_bytes{};  ///< Size in bytes
-  bool head{};          ///< Indicates whether ptr was allocated from the heap
-};
+template <typename BlockType>
+inline bool is_valid(BlockType const& b)
+{
+  return false;
+}
 
 /**
- * @brief Comparator for block types based on pointer address.
+ * @brief Prints a block (for debugging).
+ *
+ * This function template should be specialized for any type of block used by a concrete
+ * implementation of `free_list`.
  *
- * This comparator allows searching associative containers of blocks by pointer rather than
- * having to search by the contained type. Saves potentially error-prone temporary construction of
- * a block when you just want to search by pointer.
+ * @param b The block to print
  */
-template <typename block_type>
-struct compare_blocks {
-  // is_transparent (C++14 feature) allows search key type for set<block_type>::find()
-  using is_transparent = void;
-
-  bool operator()(block_type const& lhs, block_type const& rhs) const { return lhs < rhs; }
-  bool operator()(char const* ptr, block_type const& rhs) const { return ptr < rhs.pointer(); }
-  bool operator()(block_type const& lhs, char const* ptr) const { return lhs.pointer() < ptr; };
-};
+template <typename BlockType>
+inline void print(BlockType const& b)
+{
+}
 
 /**
- * @brief An ordered list of free memory blocks that coalesces contiguous blocks on insertion.
+ * @brief Base class defining an interface for a list of free memory blocks.
+ *
+ * Uses of derived classes assume the following additional methods are implemented (see
```

Comment:
I initially had pure virtual functions, as we discussed. I removed them because I only need static polymorphism. I started to add CRTP, but I realized that the base class doesn't need to call these functions, which is the purpose of CRTP.

Reply:
Fixed by clarifying the docs that these are not required, just suggested derived class interface.